src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1047 @@
+		ereport(LOG, (errmsg("found task with jobid/taskid: %ld/%ld",
+							 task->jobid, task->taskid)));
+
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		task->status = BACKGROUND_TASK_STATUS_RUNNING;
+		SET_NULLABLE_FIELD(task, pid, pid);
+
+		/* Update task status to indicate it is running */
+		UpdateBackgroundTask(task);
+		UpdateBackgroundJob(task->jobid);
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		MemoryContextSwitchTo(perTaskContext);
+
+		bool hadError = false;
+		bool cancelled = false;
+
+		/*
+		 * We reset the old message (if present). This will only retain the last message
+		 * in the catalog. Otherwise it would concatenate all retries.
+		 */
+		if (task->message)
+		{
+			pfree(task->message);
+		}
+		task->message = NULL;
+
+		while (GetBackgroundWorkerPid(handle, &pid) != BGWH_STOPPED)
+		{
+			int latchFlags = WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH;
+			int rc = WaitLatch(MyLatch, latchFlags, (long) 1000, PG_WAIT_EXTENSION);
+
+			/* emergency bailout if postmaster has died */
+			if (rc & WL_POSTMASTER_DEATH)
+			{
+				proc_exit(1);
+			}
+
+			if (rc & WL_LATCH_SET)
+			{
+				ResetLatch(MyLatch);
+				CHECK_FOR_INTERRUPTS();
+			}
+
+			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
+										  dsm_segment_address(seg));
+			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+
+			ConsumeTaskWorkerOutput(responseq, task, &hadError, &cancelled);
+
+			shm_mq_detach(responseq);
+		}
+
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		{
+			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
+										  dsm_segment_address(seg));
+			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+
+			ConsumeTaskWorkerOutput(responseq, task, &hadError, &cancelled);
+
+			shm_mq_detach(responseq);
+		}
+
+		UNSET_NULLABLE_FIELD(task, pid);
+		task->status = BACKGROUND_TASK_STATUS_DONE;
+		if (hadError)
+		{
+			/*
+			 * When we had an error we need to decide if we want to retry (keep the
+			 * scheduled state), or move to failed state
+			 */
+			if (cancelled)
+			{
+				task->status = BACKGROUND_TASK_STATUS_CANCELLED;
+			}
+			else if (!task->retry_count)
+			{
+				SET_NULLABLE_FIELD(task, retry_count, 1);
+			}
+			else
+			{
+				(*task->retry_count)++;
+			}
+
+			if (task->retry_count)
+			{
+				if (*task->retry_count >= 3)
+				{
+					/* fail after 3 retries */
+					task->status = BACKGROUND_TASK_STATUS_ERROR;
+					UNSET_NULLABLE_FIELD(task, not_before);
+				}
+				else
+				{
+					/* TODO figure out good backoff strategy, we retry in 10*retry minutes */
+					TimestampTz notBefore = TimestampTzPlusMilliseconds(
+						GetCurrentTimestamp(),
+						*(task->retry_count) * 10 * 60 * 1000);
+					SET_NULLABLE_FIELD(task, not_before, notBefore);
```

Comment:
I think the initial retry of 10 minutes might be too large. I expect most failures to be transient and caused by networking.

Can we go for a pattern of first retry after 1 minute, second in 10 minutes, and the final one after 1 hour? Perhaps it would be also good to already include a jitter for the retry, so that if we end up running parallel jobs - they won't initiate retries at the same time, potentially hammering an external service/node.

Some of our docs on the exponential backoff pattern:

- https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/implement-retries-exponential-backoff
- https://docs.microsoft.com/en-us/azure/architecture/patterns/retry

Reply:
I can change the intervals quite easily.

I am unconvinced we need jitter and to implement jitter correctly we would need to give up on the guaranteed delays of 10min and 60 minutes.

Jitter is mostly useful when connecting to an external service that can be hammered from many clients having noticed an endpoint failure at the same time. They should jitter their retries to _not_ hit the service at set intervals.

We will only have one rebalance job running at once. Meaning there is no potential for thundering herd here. In the future we control the parallelism of the executions to prevent overload after failure in a different and more controlled manner.

If we would implement Jitter here it would need to choose random times between zero and the max wait time. Which could cause a rebalance to quickly exhaust its retries for no good reason.

Reply:
Re-opening this for more discussion on the intervals. Personally, I think a 10 minute interval is already quite long and 1 hour seems like way too long. To me the maximum useful wait time in between attempts would be at most 5 minutes, and probably ~1 minute would be even better. 

I see a few scenarios where retries are useful (for shard moves):
1. The move never started because of some intermittent issue. 
2. The move failed halfway through
3. There is some permanent error, that requires an admin to change something in their DB. But after fixing that the moves automatically continue.

I don't think it's a good user experience to make nothing happen for 1 hour in any of those cases. It seems a much nicer user experience to fail a few times in a row, than to make the user wait for a very long time after they already know the problem is fixed. Personally, I would restart the rebalance manually if it was over paused for an hour.

So, my proposal is to have the following retries:
1. 1 times 5s
2. 1 times 20s
3. 30 times 1m

I agree that jitter isn't necessary at this point, but would like to point out that in general the jitter is not between zero and wait time, but usually between wait_time - 25% and wait_time + 25% (or some other %).

Reply:
@JelteF I have nothing against the proposed 5s/20s/1m slots. The 1m/10m/60m proposal was in anticipation of using this mechanism for other task types, that might include hitting external services. Since then however, we discussed having custom intervals per task type and I agree that shorter intervals in the case of the rebalancer would make for a much nicer user experience.

Reply:
This would imply that we need to set the background task queue interval to a value between 1s and 5s. Currently it is at 60s which would make each backoff implicitly 1m minimum, and possibly a 1m backoff would result in 2 intervals being done before started.

Are we ok with checking the tasks every 5 seconds? (this could cause the first backoff to hit 2 intervals - if we want to solve that we need to half the time to 2.5s I guess).

Reply:
I think checking tasks every 5 seconds is fine. It will also reduce the time it takes before `citus_rebalance_start` actually starts doing anything. Every 2 seconds seems fine too, that's what we do for the deadlock detector by default and that seems like a much heavier check.